Bug fixes related to AxisVisual

astrofrog opened the pull request:
This fixes a few bugs I ran into while using ``AxisVisual``:

* Changing the ``.rotation`` on a ``TextVisual`` didn't previously update the actual shader attribute for rotation. This became apparent when rotating the axes in 3-d, as the labels didn't follow the rotation.
* The calculation of the angle for the axis label didn't quite work when projection is present
* Warnings were emitted when the min/max for axes were the same

It's not clear to me how to properly test the first two but I think I could add a test for the last one if needed.

djhoese commented:
Thanks @astrofrog. Any idea if these might help with the tick labels being incorrect (see #1742)? CC @pcpatfire

astrofrog commented:
@djhoese - I don't **think** so. I've just added a regression test for the third bug.

djhoese commented:
At least one test environment is failing from linting issues. Maybe we need to start using black and/or a bot like stickler to do this on the github side instead of waiting for test environments.

djhoese commented on vispy/visuals/axis.py:
```diff
@@ -171,7 +171,9 @@ def _prepare_draw(self, view):
             tr = self._line.get_transform(map_from='visual', map_to='canvas')
-            x1, y1, x2, y2 = tr.map(self.pos)[:, :2].ravel()
+            trpos = tr.map(self.pos)
+            trpos /= trpos[:, 3:]
```
What is this line actually doing?

astrofrog replied:
This is the fix that is required for the resulting display coordinates to be correct in the presence of non-zero FOV in the turntable camera. I wrote this a few years ago so I don't remember how exactly it works 🤔

djhoese replied:
I'm trying to figure out what the `3:` is for. If `trpos` is (N, 4)` then I suppose this is dividing `x, y, z` coordinates by a `w` coordinate? I feel like I'm missing something obvious.

astrofrog replied:
Yes, we're basically renormalizing the coordinates based on this last fourth coordinate - I could have written it ``trpos[:, 3]`` instead of ``3:``. I can't remember the details of this but when there are projection effects the last ``w`` coordinate ends up being != 1.

astrofrog replied:
@djhoese - ah I found this old issue that discusses it! https://github.com/vispy/vispy/issues/1198#issuecomment-212790199

djhoese replied:
Ah, so the answer is math. Got it. (ok, fine, homogeneous coordinates)

djhoese commented on vispy/visuals/tests/test_axis.py:
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
```
Could you copy the header from the other modules and add it here?

astrofrog replied:
Done

djhoese commented:
I think I'm willing to pass on tests for the first two bullet points. We don't currently have tests (that I know of) that do complex things like changing the rotation of a TextVisual and comparing the results. We could have an expected image that we compare against, but I'm not a huge fan of static image tests. You could mock the update call (using the mock `wraps` keyword argument to keep the original functionality of the call) and check that it is called at least once when the rotation is changed. Thoughts?

Is the rotation calculation something that could/should move to the transform object itself? Then a test for it would be a simple unit test.

astrofrog commented:
@djhoese - I added a couple of tests:

* For the rotation calculation, I don't think it makes sense for this to live on the transform, but I split it out into a dedicated property on ``Axis`` (``_rotation_angle``) which makes the code cleaner and easier to test (I added a unit test to make sure the homogeneous coordinate normalization works properly)

* For the updating of the rotation I realized that I could just check the OpenGL attribute ``a_rotation`` so I've done that and it seems to work nicely.

So I think everything here now has a test in some form or other now.

djhoese commented:
The failing test should be "fixed" on master by dropping py3.8 wheel builds. Merging this since it isn't the thing triggering the failures. Thanks again @astrofrog!
